ftdc/ftdc.go

```diff
@@ -355,23 +369,95 @@ func (ftdc *FTDC) writeDatum(datum datum) error {
 // getWriter returns an io.Writer xor error for writing schema/data information. `getWriter` is only
 // expected to be called by `newDatum`.
 func (ftdc *FTDC) getWriter() (io.Writer, error) {
-	if ftdc.outputWriter != nil {
+	// If we have an `outputWriter` without a `currOutputFile`, it means ftdc was constructed with
+	// an explicit writer. We will use the passed in writer for all operations.
+	if ftdc.outputWriter != nil && ftdc.currOutputFile == nil {
 		return ftdc.outputWriter, nil
 	}
 
+	// Note to readers, until this function starts mutating `outputWriter` and `currOutputFile`, you
+	// can safely assume:
+	//
+	//   `outputWriter == nil if and only if currOutputFile == nil`.
+	//
+	// In case that helps reading the following logic.
+
+	// If we have an active outputWriter and we have not exceeded our FTDC file rotation quota, we
+	// can just return.
+	if ftdc.outputWriter != nil && ftdc.bytesWrittenCounter.count < ftdc.maxFileSizeBytes {
+		return ftdc.outputWriter, nil
+	}
+
+	// If we're in the logic branch where we have exceeded our FTDC file rotation quota, we first
+	// close the `currOutputFile`.
+	if ftdc.currOutputFile != nil {
+		// Dan: An error closing a file (any resource for that matter) is not an error. I will die
+		// on that hill.
+		utils.UncheckedError(ftdc.currOutputFile.Close())
+	}
+
 	var err error
-	ftdc.currOutputFile, err = os.Create("./viam-server.ftdc")
-	if err != nil {
+	// It's unclear in what circumstance we'd expect creating a new file to fail. Try 5 times for no
+	// good reason before giving up entirely and shutting down FTDC.
+	for numTries := 0; numTries < 5; numTries++ {
+		now := time.Now().UTC()
+		// lint wants 0o600 file permissions. We don't expect the unix user someone is ssh'ed in as
```

Comment:
```suggestion
		// lint wants 0o644 file permissions. We don't expect the unix user someone is ssh'ed in as
```

Reply:
The existing documentation is intended. We are making a call with 644 as the permissions setting. And that is* intended. But the linter complains. The linter says the permissions should be 600. Such that only the current unix user can read or write to the file.

Reply:
😮‍💨 cool that makes sense; should have read closer.